### ISSUE TITLE
`IconButton` should include `Tooltip` by default

### ISSUE THREAD

**Issue by iansan5653** [primer/react#2008]
## Summary

There is no documented example of how to render an icon-only button with an accessible tooltip in a way that would meet all of the following standard requirements:

- Render as an icon-only button with a single focus target
- Read the label one time in a manner that ties it to the button, for screen reader users
- Show the label on hover, for mouse users
- Show the label on focus, for keyboard users

If this is a supported behavior, then some documentation is needed to show how to do it. If this is not supported, it would be very useful functionality to add to Primer.

The [`IconButton`](https://primer.style/react/IconButton) documentation provides an example of how to label the component in an accessible manner:
```tsx
<IconButton aria-label="Search" icon={SearchIcon} />
```

And the [`Tooltip`](https://primer.style/react/Tooltip) documentation has the same:
```tsx
<Tooltip aria-label="Hello, Tooltip!">Text with a tooltip</Tooltip>
```

Both of these examples use `aria-label` however, which provides no obvious way to create an icon button that shows its accessible label on hover as a tooltip. 

## Use Cases

This is a very common requirement. See for example these instances on github.com:

**Notifications bell**:
<img width="391" alt="Screenshot of the GitHub notifications bell button showing a tooltip that says 'you have unread notifications'" src="https://user-images.githubusercontent.com/2294248/161074778-dd755940-5cbe-440a-8712-840227937c52.png">

**Markdown editor toolbar**:
<img width="192" alt="Screenshot of the GitHub markdown editor 'bold' button showing a tooltip" src="https://user-images.githubusercontent.com/2294248/161074856-157f1b16-4023-48e6-8b18-6e59275d408d.png">

**Code view**:
<img width="192" alt="Screenshot of the GitHub pull request code view, showing the 'Expand all' button with a tooltip" src="https://user-images.githubusercontent.com/2294248/161075094-686ed8d6-b6a0-4fd1-aa76-a0bedee80a60.png">

This is recommended as a best practice by [Primer's own tooltip guide](https://primer.style/design/accessibility/tooltip-alternatives):

> <img width="473" alt="Do: Use a tooltip to give a text label to an icon button" src="https://user-images.githubusercontent.com/2294248/162039835-56f30f63-0ea9-4780-9bf0-fce995c2a964.png">

## Attempts

I've tried all of the following examples but I haven't been able to find a satisfying way to do this:

**Label on tooltip**: Tooltip does not appear on button focus and the button appears as unlabelled in the accessibility tree.
```tsx
<Tooltip aria-label="Bold">
  <IconButton icon={BoldIcon} />
</Tooltip>
```

**Label on button only**: Tooltip does not work at all.
```tsx
<Tooltip>
  <IconButton icon={BoldIcon} aria-label="Bold" />
</Tooltip>
```

**Label on both**: The label will be read twice to screen readers (something like "Tooltip: Bold, Button: Bold") which is unnecessarily redundant. Also the tooltip still doesn't appear on focus.
```tsx
<Tooltip aria-label="Bold">
  <IconButton icon={BoldIcon} aria-label="Bold" />
</Tooltip>
```

**Tooltip as button**: Seems like a promising idea but `Tooltip` does not support the `as` prop.
```tsx
<Tooltip aria-label="Bold" as={IconButton} icon={BoldIcon} />
```

**Button as tooltip**: Renders an unfocuseable `span` instead of a `button`.
```tsx
<IconButton icon={BoldIcon} as={Tooltip} aria-label="Bold" />
```

**Tooltip in button children**: Does not work because the `IconButton` ignores children.
```tsx
<IconButton icon={BoldIcon} aria-label="Bold">
  <Tooltip aria-label="Bold" role="presentation">Bold</Tooltip>
</IconButton>
```

**Tooltip as button icon**: This is horribly inelegant, but it does generate a good accessibility tree and still shows the tooltip on hover. It unfortunately still doesn't show the tooltip on focus, but it's the best I have for now.
```tsx
<IconButton
  icon={() => <Tooltip aria-label="Bold" role="presentation"><BoldIcon /></Tooltip>}
  aria-label="Bold"
/>
```

## Proposed Solution

My ideal solution would be to just show the `IconButton`'s `aria-label` as a tooltip by default.

To configure this, I propose adding two new props to the `IconButton` component:

- `disableTooltip`: Turn off the tooltip.
- `tooltipProps`: Configure the tooltip by passing props to the underlying `Tooltip` component.

Then I could achieve my goal simply with:
```tsx
<IconButton icon={BoldIcon} aria-label="Bold" />
```

**Comment by siddharthkp** [comment]
> My ideal solution would be to just show the IconButton's aria-label as a tooltip by default

+1 to that. I like the idea of having a tooltip always.

`disableTooltip`: What are the instances where you would want to disable the tooltip?

We'd need to workshop `tooltipProps` a bit before committing to an API 🤔

**Comment by iansan5653** [comment]
> What are the instances where you would want to disable the tooltip?

I figured there might be a case where the built-in tooltip conflicts with another tooltip or hover effect, but I can't think of a concrete example. Maybe it's actually not necessary to support disabling it.

**Comment by lesliecdubs** [comment]
I'm going to run this idea of having a tooltip by default at tomorrow's Primer Patterns working group session to get any feedback or concerns before we run with that idea.

In the meantime - @khiga8 could you please take a quick look at this issue and the proposal to adding a tooltip-by-default on `IconButton` to ensure this approach is in line with our accessibility needs?

**Comment by khiga8** [comment]
Hi @lesliecdubs!

@hectahertz can speak more to the work on `IconButton` in PVC! I believe we were leaning towards having tooltip by default for all `IconButton` for accessibility purposes. Discussion can be found in: https://github.com/primer/view_components/pull/1062.

As a side note, I notice in all the examples listed in this issue, the tooltip is always semantically associated as a label. I'm not sure if this is out of scope, but in the long run, the tooltip should also be allowed to be associated to the icon button as a supplementary description. 

You can check out the markup of the markdown toolbar in dotcom as an example. These were recently updated to use tooltips that are associated semantically using `aria-describedby` with `aria-label` directly set on the button with a brief accessible name. This was done (at least in dotcom) because the tooltip content is kind of long and contains keyboard shortcuts which makes the tooltip inappropriate as an accessible name. Accessible names should be brief. Therefore you'll notice we have a label like `Bold` directly on the button and a tooltip associated as description that contains, `Add bold text, <Cmd+b>`. 

We don't have this in dotcom, but similarly in the `Notifications` button example, it may be appropriate to set `aria-label` as `Notifications`, and set a tooltip as supplementary description with `You have unread notifications`.

We've recently worked on a bunch of accessibility doc improvements on the PVC and Primer Design side to encourage proper tooltip usage with valid tooltip use examples. Here are some references that may be helpful as Primer React team works on tooltip:

- [PVC: Tooltip](https://primer.style/view-components/components/alpha/tooltip#as-a-description-for-an-icon-only-button)
- [Tooltip alternatives](https://primer.style/design/accessibility/tooltip-alternatives)

**Comment by lesliecdubs** [comment]
Thank you for weighing in @khiga8, this is very helpful context to inform our next steps!

It does indeed look like PVC came to the same decision about always showing a Tooltip on Icon Button in https://github.com/primer/view_components/pull/1062:
> After https://github.com/primer/view_components/pull/1062#issuecomment-1071065360, we decided to always include a Tooltip on IconButton. The aria-label` property is always required, and this will improve the experience for sighted users. Hopefully, this approach will simplify the API and will keep the instances of IconButton consistent and accessible.

@hectahertz can you weigh in on any other API decisions that have been made in PVC that may be relevant? Perhaps @hectahertz and @siddharthkp can sync in these comments on whether we want to provide support for props `disableTooltip` or `tooltipProps` as proposed in this issue, and whether there are other specific learnings from the PVC work that should inform PRC's approach.

**Comment by siddharthkp** [comment]
From: https://github.com/primer/view_components/pull/1062, IconButton has an [argument called `tooltip_direction`](https://view-components-drq813o3o-primer.vercel.app/view-components/components/iconbutton#:~:text=for%20more%20information.-,tooltip_direction,-Symbol) which is the only customisation tooltip allows. ([tooltip arguments](https://view-components-drq813o3o-primer.vercel.app/view-components/components/alpha/tooltip#arguments))

@hectahertz @khiga8 Are there plans on expanding this list of arguments?

> As a side note, I notice in all the examples listed in this issue, the tooltip is always semantically associated as a label. I'm not sure if this is out of scope, but in the long run, the tooltip should also be allowed to be associated to the icon button as a supplementary description.

Thanks for the example, this is really helpful! It was one of the questions I had :)

**Comment by siddharthkp** [comment]
tl;dr: we should copy what @hectahertz did in https://github.com/primer/view_components/pull/1062

The [Tooltip in primer/react](https://primer.style/react/Tooltip#component-props) is more customisable, so the API could look end up looking slightly different:

Here are 2 use cases: 

1. IconButton with an `aria-label`:

    <img width="293" alt="image" src="https://user-images.githubusercontent.com/1863771/161742401-a744aec2-7cc0-4f9d-9de0-505a4587e1fe.png">
    
    Without tooltip:
    
    ```jsx
    <IconButton icon={KebabIcon} aria-label="More options" />
    ```
    
    &nbsp;
    
    Option 1: wrap IconButton in a Tooltip
    
    ```jsx
    <Tooltip text="More options" direction="sw"  align="left">
      <IconButton icon={KebabIcon} /> // need to make aria-label optional
    </Tooltip>
    ```
     👍 If you've used the Tooltip component before, composing it this way is intuitive
     👍 Tooltip attributes go on the tooltip as you would expected
     
     👎 Need to make `aria-label` on `IconButton` optional so that it is not repeated twice. Will result in IconButton without a Tooltip*

    &nbsp;
    
    Option 2: bake in tooltip with IconButton (use aria-label used for tooltip text)
    
    ```jsx
    <IconButton icon={KebabIcon} aria-label="More options" tooltipDirection="sw" tooltipAlign="left" />
    // or
    <IconButton icon={KebabIcon} aria-label="More options" tooltipProps={{direction: 'sw', align: 'left'}} />
    ```
    👍 Can ensure there will always be a tooltip with IconButton

    👎 tooltipProps on the IconButton aren't intuitive
    👎 while unintuitive, it's hopefully rare to want to customize the tooltip's position. we can attempt to restrict the options to just `tooltipDirection` to avoid the more awkward `tooltipProps` API
    
   
    While option 1 is more composable, we should bake in the tooltip in Option 2 to make the recommended path the default one.

    &nbsp;

2. For the markdown editor with `aria-label` and `aria-describedby`
    - `aria-label` is "Bold" 
    - `aria-describedby` = tooltip text is "Add bold text, <Cmd+b>"
    <img width="183" alt="image" src="https://user-images.githubusercontent.com/1863771/161740581-89794d6a-f089-4cd1-acd2-25a41d585f77.png">
    
    
    Option 1:
    
    ```jsx
    <Tooltip text="Add bold text, <Cmd+b>" direction="s">
      <IconButton icon={BoldIcon} aria-label="Bold" />
    </Tooltip>
    ```

     👍 If you've used the Tooltip component before, composing it this way is intuitive
     👍 Tooltip attributes go on the tooltip as you would expected
     👍 (compared to previous example) Don't need to make `aria-label` optional, this is perfect.
    
    Option 2:
    
    ```jsx
    <IconButton 
      icon={BoldIcon}
      aria-label="Bold"
      aria-describedby="Add bold text, <Cmd+b>"  // aria-describedby used for tooltip
      tooltipDirection="s"
    />
    ```

     👍 collocating aria-describedby right next to aria-label on the IconButton itself makes a lot of sense while reading the code (for someone who works with accessible markup in mind, this is very intuitive. _I have secondary information about this button, so that goes on aria-describedby_)
     👎 It's not immediately obvious that the tooltip will pick up aria-describedby instead of aria-label (for someone who is approaching this visually, it isn't intuitive. _I want to add a tooltip, but instead of using a Tooltip component, i should add aria-describedby because this is an IconButton_)
     👎 2 ways of doing the same thing - if someone writes the code from Option 1, that would also work and give the same exact output 😅 


&nbsp;

Personal opinion

If we look at the library as a whole and not just IconButton, it makes more sense to go with Option 1, re-use Tooltip and ask folks to wrap IconButton with a Tooltip. It's easy to miss, but we can enforce this with help of our linter or/and by adding warnings in dev mode. [example of warning, sandbox](https://codesandbox.io/s/admiring-panna-tbzvsi?file=/src/App.tsx:379-453)

However, I think being _accessible out of the box_ makes Option 2 very attractive despite the awkwardness around customising tooltip position. If we know developers would always have to wrap an IconButton with a tooltip, we should do it for them.

Having 2 ways of doing one thing from the markdown example feels wrong, but is an _okay_ tradeoff to make as long as the output for both approaches is accessible.

**Comment by vdepizzol** [comment]
> I'm going to run this idea of having a tooltip by default at tomorrow's Primer Patterns working group session to get any feedback or concerns before we run with that idea.

@lesliecdubs thank you for bringing this issue to the Primer Patterns working group attention 👋 

@siddharthkp said:
> (...) I think being accessible out of the box makes Option 2 very attractive despite the awkwardness around customising tooltip position. If we know developers would always have to wrap an IconButton with a tooltip, we should do it for them.

I agree we should provide components that are accessible by default.

A few considerations:

### `aria-label` + `aria-describedby`

I wonder if we could abstract the `aria-label` into a more generic `label` prop for IconButton, while still generating `aria-label` + `aria-describedby` at the end. That way, it's not that `aria-label` also gets rendered as a tooltip, but that the IconButton `label` is initially used both as an `aria-label`, and as a tooltip:

```jsx
<IconButton
  icon={HeartIcon}
  label="Sponsor"
/>
```

`aria-describedby`, as the [HTML attribute specifies](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-describedby), asks for an element ID. In this case, we're passing a string directly. That could cause some confusion. Maybe we could call it `tooltipText` or `tooltipDescription`, to make it obvious it's a variation of the label specifically for the tooltip text.

```jsx
<IconButton
  icon={BoldIcon}
  label="Bold"
  tooltipText="Add bold text"
/>
```

### Optional tooltips

Although tooltips should come by default and be present in most IconButton components, it might make sense to decouple the tooltip behavior if necessary.

I can imagine larger components using IconButton for convenience (such as expanders, split buttons, etc) which wouldn't need a tooltip based on their context (but would need an `aria-label`, of course!).

An example I can think about is the [TreeGrid](https://github.com/github/primer/issues/310#issuecomment-1088315302) I'm currently working on. The placement and context of the icon are enough for describing behavior. In that case we could have something like `disableTooltip`:

```jsx
<IconButton
  icon={ChevronRightIcon}
  label="Expand folder"
  disableTooltip
/>
```

That could also be useful when a custom rendering is used, or when there's a need for a custom tooltip with extra parameters:

```jsx
<Tooltip text="Add bold text" direction="s">
  <IconButton
    icon={HeartIcon}
    label="Sponsor"
    disableTooltip
  />
</Tooltip>
```

Or, if an IconButton is a child of Tooltip, could we enforce a `disableTooltip` directly?

```jsx
<Tooltip text="Add bold text" direction="s">
  <IconButton
    icon={HeartIcon}
    label="Sponsor"
  />
</Tooltip>
```

Note that the `label` prop remains required.

### Shortcuts

I think we should have a better way to reference shortcuts in the tooltips, instead of bundling them together with the description.

@maximedegreve has done some [explorations](https://www.figma.com/file/VgFQMEiZ5clZK60ipdYXc6/Restyle-%3Ckbd%3E?node-id=101%3A29338) for improved `<kbd>`, which I think could be brought into tooltips as well:

![ActionMenu with new shortcut styles](https://user-images.githubusercontent.com/293280/161866773-37ec2d6d-5ca0-493e-80a8-34c12561836f.png)

[Here's a quick test](https://www.figma.com/file/k5j7kueli3GfMxkierLRtf/primer%2Freact%232008?node-id=0%3A1):

<img src="https://user-images.githubusercontent.com/293280/161866724-d2e061e6-2605-428a-8807-589735f42375.png" alt="Tooltips with dedicated shortcut visuals" width="220" />

An ideal scenario could have a Tooltip prop with common keys (`Command`, `Shift`, etc) automatically switching to their correct glyphs, and supporting multiple OSs:

```jsx
<IconButton
  icon={ChevronRightIcon}
  label="More actions"
  shortcutPreview={{mac: 'Command + Shift + M', win: 'Ctrl + Shift + M'}}
/>
```

In the example above I'm using `shortcutPreview` to make sure the prop doesn't make it sound like it's actually adding a shortcut to the action.

@siddharthkp let me know your thoughts! 🙌

**Comment by iansan5653** [comment]
I like 'Option 3: Opt-in composition' - it provides significant power for complex scenarios. Having `label` be always required and of type `string` ensures that the button's `aria-label` is always valid. 

I can see the last example generating a structure akin to:

```html
<div>
  <button aria-label="Bold" aria-describedby="tooltipId"><!--bold icon--></button>
  <span id="tooltipId">Add bold text, <kbd>⌘B</kbd></span>
</div>
```

From a CSS perspective, it wouldn't be too hard to have a containing `Tooltip` prevent the `aria-label` tooltip from appearing and to have the container show its tooltip on `focus-within`.

**Comment by vdepizzol** [comment]
@siddharthkp love where this is going! Option 3 makes a lot of sense.

Just a comment on shortcut placement. I believe we'd still benefit from a specific slot for shortcuts instead of placing them inline with the tooltip text. I think it'd be a future-proof solution, and it'd give us more flexibility on how to handle their rendering... Example:

```jsx
<Tooltip 
  text="Add bold text"
  shortcutPreview={<Key meta>B</Key>}
  direction="sw"
>
  <IconButton icon={BoldIcon} label="Bold" />
</Tooltip>
```

**Comment by mperrotti** [comment]
@siddharthkp - I agree that 'Option 3: Opt-in composition' is the right way forward.

Just a few observations:

**1.** Since `IconButton` renders a tooltip by default, a Primer customer could feel uncertainty when nesting it inside of a `Tooltip`

```jsx
// This will show a tooltip that reads "Notifications"
<IconButton icon={BellIcon} label="Notifications" />

// Am I supposed to pass `tooltip={false}` to prevent two tooltips from being shown?
<Tooltip text="You have no unread notifications" direction="sw">
  <IconButton icon={BellIcon} label="Notifications" tooltip={false} />
</Tooltip>
```

Did you have an idea for how we could suppress the `IconButton` tooltip if it's a child of `Tooltip`? I'm assuming we'd use React Context.

**2.** Instead of `tooltip={false}` to override the default behavior, I'd suggest we phrase it like `hideTooltip`. That way, we can just check if `hideTooltip` is `true` instead of checking if it's defined _and_ that it's false. Small bonus: users can just pass `hideTooltip` instead of `tooltip={false}`. See the `visuallyHidden` prop on [FormControl.Label](https://primer.style/react/FormControl#formcontrollabel) for another example.

**3.** I think @vdepizzol's `shortcutPreview` idea is helpful for maintaining consistent layouts, but I don't like that it's specific to keyboard shortcuts, and I think it will end up being used to render things other than keyboard shortcuts. Maybe we change it to something more generic like `trailingContent` or a child component like `Tooltip.TrailingContent`?

If we want to keep `shortcutPreview` and _only_ allow keyboard shortcuts, maybe it could accept an array of keyboard keys instead:
```jsx
<Tooltip 
  text="Add bold text"
  shortcutPreview={['meta', 'B']}
  direction="sw"
>
  <IconButton icon={BoldIcon} label="Bold" />
</Tooltip>
```

**Comment by pksjce** [comment]
Love this discussion. 

First off, I agree with @khiga8 that the label and the tooltip content must be separate entities. They both have different functions. Label is an immediate quick recognition of the button's functionality while the tooltip is an elaborate context driven description. If the tooltip text is going to be the same as the button label text, that is not an appropriate use of the tooltip.

Additionally, providing a default tooltip with the button is not a great idea. This leads to unnecessary API overload on the Button like `tooltipProps` and `hideTooltip`. We want to use something like `hideTooltip` sparingly but with this pattern it might become a norm than an escape hatch. 

My vote is for the `Tooltip` component to intentionally wrap around the `IconButton` when a tooltip is required. This can then become a consistent pattern for all `Tooltip` usage. For example, it's not uncommon to have tooltips on normal `Button` components, `Input`, `Link` components etc as shown [here](https://dequeuniversity.com/library/aria/tooltip).

So in the simple use case we have 

```
<Tooltip 
  text="Add bold text"
  direction="sw"
>
  <IconButton icon={BoldIcon} label="Bold" />
</Tooltip>
```

Further, we should have additional API for Tooltip to separate the trigger and rendered content to support more complex usecases like components in tooltips and dynamically rendered tooltips.

```
import {Key} from '@primer/react'
import * from '@primer/react/tooltip'

<Tooltip direction="sw">
  <TooltipTrigger>
    <IconButton icon={BoldIcon} label="Bold" />
  </TooltipTrigger>
  <TooltipContent>
    <>Add bold text, <Key meta>B</Key></>
  </TooltipContent>
</Tooltip>
```

**Comment by siddharthkp** [comment]
> @mperotti: Am I supposed to pass `tooltip={false}` to prevent two tooltips from being shown?

Nope, we'll disable it automatically. Both context and children magic would work here.

> @mperrotti: Instead of tooltip={false} to override the default behavior, I'd suggest we phrase it like hideTooltip.

Yep, I agree with you. `tooltip=false` only made sense while there was a `tooltip="some text"` prop on the `IconButton`. 

> @pksjce: We want to use something like hideTooltip sparingly but with this pattern it might become a norm than an escape hatch.

Yes 💯  This should only be in super rare (and hopefully internal) use cases where there is explicitly not a tooltip on the icon button (like TreeGrid)

> @pksjce: My vote is for the Tooltip component to intentionally wrap around the IconButton when a tooltip is required. This can then become a consistent pattern for all Tooltip usage.
> I agree with @khiga8 that the label and the tooltip content must be separate entities. They both have different functions.

Vote recorded! 

You're right, ideally tooltip content should be a description instead of simple label. But, I fear folks would often not use one 😢 

Personal opinion: anything we can do by default, without getting in the way of them doing better is a step in the right direction. [longer personal opinion](https://github.com/primer/react/issues/2008#:~:text=If%20we%20look,it%20for%20them.)

---

> @mperrotti: Maybe we change it to something more generic like trailingContent or a child component like Tooltip.TrailingContent?

> @pksjce: Further, we should have additional API for Tooltip to separate the trigger and rendered content to support more complex usecases like components in tooltips and dynamically rendered tooltips.


I agree with both of you. But, I think we should exclude shortcuts for this conversation and come back to it after along with other components that use shortcuts. There is a broader pattern there.

**Comment by iansan5653** [comment]
I really like where this is going. One more small thing I noticed today wrt keyboard shoprtcuts - there is an [`aria-keyshortcuts`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-keyshortcuts) attribute that should probably be set on the button.

Maybe the default tooltip content should be something like (please excuse my quick pseudocode):
```ts
<>
  {ariaLabel} {ariaKeyshortcuts.split(' ').map(shortcut =>
    shortcut.split('+').map(key => <Key>{key}</Key>)
  ).join(' / ')}
</>
```

In other words, maybe we should render the `aria-keyshortcuts` into the tooltip by default if we are already rendering the `aria-label` there.

**Comment by siddharthkp** [comment]
Putting the html here for validation:


```html
<!-- Default tooltip -->
<IconButton icon={BellIcon} aria-label="Notifications" />

<span role="tooltip" id="tooltip-id" aria-label="Notifications">
  <button aria-labelledby="tooltip-id">
    <svg></svg>
  </button>
</span>

<!-- Custom tooltip text -->
<Tooltip text="You have no unread notifications">
  <IconButton icon={BellIcon} aria-label="Notifications" {...args} />
</Tooltip>

<span role="tooltip" id="tooltip-id" aria-label="You have no unread notifications">
  <button aria-label="Notifications" aria-describedby="tooltip-id">
    <svg></svg>
  </button>
</span>
```

**Comment by siddharthkp** [comment]
Turns out our Tooltip component isn't really accessible, so that's a slight increase in scope, updated size to `rock` 😅

**Comment by siddharthkp** [comment]
Reopening because we reverted https://github.com/primer/react/pull/1984, now tackled in https://github.com/primer/react/pull/2024

**Comment by mperrotti** [comment]
This also affects the icon-only SegmentedControl buttons because they use a tooltip to show the hidden label.

Instead of creating a new a11y issue for SegmentedControl, I'm including this info here. Let me know if a new issue specific to SegmentedControl would be more helpful, and I'll create it.

**Comment by lesliecdubs** [comment]
I've moved this from Inbox back to Unprioritized Backlog for now. @siddharthkp I think everyone else is heads down right now, but if you had space to take a look at this again this quarter, we could consider moving it up next. If not, we can reassess before the next planning cycle at the end of September.

**Comment by maximedegreve** [comment]
Cross posting: https://github.com/github/primer/issues/1244

**Comment by broccolinisoup** [comment]
Hi everyone 👋🏻 I am working on remediating Tooltip accessibility issues and IconButton is a part of it. Please see the batch issue https://github.com/github/primer/issues/2137 if you are curious about the details of the overall remediations and the release strategy.

I just caught up with all the comments here and as far as I understand, the suggested PRC IconButton API for label and description type tooltip looks like below (copy-paste from earlier)

### IconButton with tooltip
```jsx
// by default, we use label to add tooltip. 
// ⭐️ accessible by default
<IconButton icon={KebabIcon} label="More options" />
```
### IconButton with tooltip text that's different than label (Supplementary text)
```jsx
// if you want to change the tooltip text or direction,
// opt-in into composition and wrap the IconButton in a tooltip component
<Tooltip text="You have no unread notifications" direction="sw" >
  <IconButton icon={BellIcon} label="Notifications" />
</Tooltip>
// good things: Tooltip props are on the Tooltip. If you've used the Tooltip before, this is very intuitive.
```

However in PVC, all comes within the box. So the equivalent of PVC API would be 

### IconButton with tooltip

```jsx
// Same except the prop name PVC uses `aria-label` versus `label`
// ⭐️ accessible by default
<IconButton icon={KebabIcon} aria-label="More options" />
```
### IconButton with tooltip text that's different than label (Supplementary text)
```jsx
  <IconButton icon={BellIcon} aria-label="Notifications" aria-description="You have no unread notifications"/>
```

I am very curious to hear what you think. Feel free to add your questions and concerns and hopefully we will be resolving this issue soon ⭐

**Comment by siddharthkp** [comment]
> ### IconButton with tooltip
> // Same except the prop name PVC uses `aria-label` versus `label`
> // ⭐️ accessible by default
> `<IconButton icon={KebabIcon} aria-label="More options" />`

perfect default case. love it, ship it :shipit: 

&nbsp;

> ### IconButton with tooltip text that's different than label (Supplementary text)
>   `<IconButton icon={BellIcon} aria-label="Notifications" aria-description="You have no unread notifications"/>`

Obviously biased, so feel free to ignore, I like my suggestion of wrapping Tooltip explicitly when you want any customisations for tooltip.

I think bringing the API from PVC has some issues:

```tsx
<IconButton
  icon={BellIcon}
  aria-label="Notifications"
  aria-description="You have no unread notifications"
/>
```

1. API's not very intuitive: It's not obvious looking at the API that if you want to customise tooltip text, you should add `aria-description`. It's something you _have to know_ or find in the docs, not something you can guess.

   While I like the idea of using aria-props in the API when applicable, I feel like we are misusing the prop here because `aria-description` doesn't actually go on the button in PVC. Instead, we use it as visible text in the tooltip and add `aria-describedby` to the button for screen readers.
   
   ```tsx
   // API:
   <IconButton
     icon={BellIcon}
     aria-label="Notifications"
     aria-description={<>Add bold text, <Key meta>B</Key></>}
   />
   
   // vs rendered: 
   <div class="position-relative">
      <button
        aria-label="Bold"
        id="icon-button-hashed"
        aria-describedby="tooltip-hashed"
      >
        <svg aria-hidden="true"></svg>
      </button>
      <tool-tip
        role="tooltip"
        id="tooltip-hashed"
        for="icon-button-hashed"
        data-type="description"
      >
        Add bold text, Cmd+b
      </tool-tip>
   </div>
   ```
   
   Another use case is when the tooltip can have rich text (not interactive)  inside it, which would be invalid for `aria-description` but are valid for tooltip text. Asking to use rich text in `aria-description` because it's not actually `aria-description` feels unintuitive. (example from memex, need to validate if this is a use case we should support or is it being used to create inaccessible ui)
   
    ```tsx
   <IconButton
     icon={BellIcon}
     aria-label="Notifications"
     aria-description={<>Add bold text, <Key meta>B</Key></>}
   />
   ```
   
2. Tooltip props need to go on the IconButton: An easy way to solve the above issue would be to hide implementation detail and accept a tooltip specific prop like `tooltip` or `tooltipText` instead.

   ```tsx
   <IconButton
     icon={BellIcon}
     aria-label="Notifications"
     tooltip="Add bold text, Cmd+b"
   />
   ```

   It's not pretty, but it's totally fine if it's just one prop.
   
   But it's probably not just one prop, if you want to customise the direction of the tooltip, you need to accept another prop on IconButton which looks something like `tooltip_direction` from PVC.

   If we want to allow style customisation, we need to accept props like `tooltipSx` or `tooltipClassName` to passthrough as well, which feels like the props are on the wrong place. (note: tooltip is a sibling of button, not child, it's not possible to target the tooltip from button's [sx](https://musical-adventure-wlr3n3k.pages.github.io/?name=tooltip&attribute=sx&repo=github/github) / [classname](https://musical-adventure-wlr3n3k.pages.github.io/?name=tooltip&attribute=classname&repo=github/github))
  
   ```tsx
   <IconButton
     icon={BellIcon}
     aria-label="Notifications"
     tooltipText="You have no unread notifications"
     tooltipDirection="sw"
     tooltipClassName="icon-button-with-tooltip__tooltip"  // from memex
   />
   
   // or 
   
   ```tsx
   <IconButton
     icon={BellIcon}
     aria-label="Notifications"
     tooltipProps={{
       text="You have no unread notifications"
       direction="sw"
       className="icon-button-with-tooltip__tooltip"  // from memex
     }}
   />
   ```
   
&nbsp;

### Suggestion: Opt-in composition

**1. Default happy case**

```tsx
// by default, we use label to add tooltip. 
// ⭐️ accessible by default
<IconButton icon={KebabIcon} label="More options" />
```

**2. IconButton with tooltip text that's different than label (Supplementary text)**: If you want to customise the tooltip text, direction or styles, you can opt-in to composition and wrap the IconButton in a tooltip component

```tsx
<Tooltip text="You have no unread notifications" direction="sw"  className="icon-button-tooltip">
  <IconButton icon={BellIcon} label="Notifications" />
</Tooltip>
```

👍  Tooltip props are on the Tooltip. 
👍  Looks like other [usage of tooltip](https://primer.style/react/Tooltip#examples), consistent is intuitive

Also works for rich content (if we want to allow it):

```tsx
<Tooltip 
  text={<>Add bold text, <Key meta>B</Key></>}
  direction="sw"
>
  <IconButton icon={BoldIcon} label="Bold" />
</Tooltip>
```


### Different API for PVC and PRC?

I think there's a case to bring these suggestions to PVC **if we like this API more**, but I personally think it's alright to have some API differences between PVC and PRC based on what feels intuitive in the mental model of Rails and React.

**Comment by joshblack** [comment]
I feel like a great point that you bring up @iansan5653 is that it'd be great if we can make illegal (or non-ideal) states unrepresentable. I think you had a great example with:

```tsx
<Tooltip text="Add bold text" keyboardShortcut="Alt+B">
  <IconButton icon={BoldIcon} label="Bold" keyboardShortcut="CommandOrControl+B" />
</IconButton>
```

To me this demonstrates a challenge that can come up where you have ambiguity as to what will happen in the composition case (specifically what prop value "wins" here). I do think that over time you can infer that this is a merge between the two and that the inner one wins out, ultimately.

I think an advantage of `IconButton` owning the props in this scenario is that it is able to control the set of permissible values provided to `Tooltip` that make sense when used alongside `IconButton`.

Either way, curious what you would recommend @broccolinisoup with all the work that you've been doing to see what would help make this the easiest change on your end 😅 lol

**Comment by broccolinisoup** [comment]
👋🏻 Here is the API proposal for the IconButton 🎉 https://github.com/github/primer/pull/2339

We left the keyboard shortcut stuff to deal with later and tried addressing more prominent issues in the first iteration. Other than that, I think we addressed all issues we talked here. Let me know if I miss anything and I am looking forward to hearing your feedback on the PR 🙌🏻

**Comment by lindseywild** [comment]
@broccolinisoup Hi! Is it accurate to say that this issue will be fixed when https://github.com/github/primer/issues/1244#issuecomment-2041912201 is completed?
